rbd/encryption_load2.go

```diff
@@ -0,0 +1,63 @@
+//go:build !octopus && !pacific && !quincy && ceph_preview
+
+package rbd
+
+// #cgo LDFLAGS: -lrbd
+// /* force XSI-complaint strerror_r() */
+// #define _POSIX_C_SOURCE 200112L
+// #undef _GNU_SOURCE
+// #include <rbd/librbd.h>
+import "C"
+
+import (
+	"unsafe"
+)
+
+// toEncryptionSpec returns a rbd_encryption_spec_t converted from the
+// cEncryptionData type.
+func (edata cEncryptionData) toEncryptionSpec() C.rbd_encryption_spec_t {
+	var cSpec C.rbd_encryption_spec_t
+	cSpec.format = edata.format
+	cSpec.opts = edata.opts
+	cSpec.opts_size = edata.optsSize
+	return cSpec
+}
+
+// EncryptionLoad2 enables IO on an open encrypted image and ancestor images.
+// The first EncryptionOptions in the slice is applied to the image, the second
+// to the first ancestor, the third to the second ancestor and so on.
+// If the length of the slice is smaller than the number of ancestors the
+// final item in the slice will be applied to all remaining ancestors.
+//
+// Implements:
+//
+//	int rbd_encryption_load2(rbd_image_t image,
+//	                         const rbd_encryption_spec_t *specs,
+//	                         size_t spec_count);
+func (image *Image) EncryptionLoad2(opts []EncryptionOptions) error {
+	if image.image == nil {
+		return ErrImageNotOpen
+	}
+
+	length := len(opts)
+	eos := make([]cEncryptionData, length)
+	cspecs := (*C.rbd_encryption_spec_t)(C.malloc(
+		C.size_t(C.sizeof_rbd_encryption_spec_t * length)))
+	specs := unsafe.Slice(cspecs, length)
+
+	for idx, option := range opts {
+		eos[idx] = option.allocateEncryptionOptions()
```

Comment:
Together with `rbd_encryption_load2()`, librbd also introduced `RBD_ENCRYPTION_FORMAT_LUKS` and a respective options struct.  This is a wildcard format suitable only for loading encryption (i.e. not for `rbd_encryption_format()` which must format the image to a specific format), intended to allow the user to load either LUKS1 or LUKS2 without explicitly specifying which.
```
typedef enum {
    RBD_ENCRYPTION_FORMAT_LUKS1 = 0,
    RBD_ENCRYPTION_FORMAT_LUKS2 = 1,
    RBD_ENCRYPTION_FORMAT_LUKS  = 2
} rbd_encryption_format_t;
```
```
typedef struct {
    const char* passphrase;
    size_t passphrase_size;
} rbd_encryption_luks_format_options_t;
```
It would be good to expose that as well.

Reply:
done